File: algorithm_datastructure/leetcode/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#

# @lc code=start
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        length = len(nums)
        if length == 0:
            return [-1, -1]

        # Left close, right close
        left = 0
        right = length - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                # ! nums[mid] == target!
                # The right is included in the results !!!
                # Find the min and max in this range !!!
                for i in range(left, right + 1):
                    if nums[i] == target:
                        # Avoid count left twice
                        if nums[left] != target:
                            left = i
                        right = i
                return [left, right]

        # Not found!
        return [-1, -1]


# @lc code=end

File: algorithm_data_structure/leetcode/34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

int binarySearch(vector<int> &nums, int left, int right, int target)
{

    while (left <= right)
    {
        int mid = left + (right - left) / 2;
        if (nums[mid] == target)
        {
            return mid;
        }
        if (target > nums[mid])
        {
            left = mid + 1;
        }
        if (target < nums[mid])
        {
            right = mid - 1;
        }
    }
    return -1; // Not found !
}

class Solution
{
public:
    vector<int> searchRange(vector<int> &nums, int target)
    {
        vector<int> ret;
        int left = -1, right = -1;
        int pos = binarySearch(nums, 0, nums.size() - 1, target);

        if (pos >= 0)
        {
            left = right = pos;

            // Search the range on the left of pos
            int l = left;
            do
            {
                left = l;
                l = binarySearch(nums, 0, left - 1, target);
            } while (l >= 0);

            // Search the range on the right of the pos.
            int r = right;
            do
            {
                right = r;
                r = binarySearch(nums, right + 1, nums.size() - 1, target);
            } while (r >= 0);
        }

        ret.push_back(left);
        ret.push_back(right);

        return ret;
    }
};
// @lc code=end
